### PR TITLE
first stab at supporting more of LGR

### PR DESCRIPTION
**Pull request by xrotwang**

### CODE REVIEW

**Comment by xrotwang** [comment]
@fmatter Does this look like what you had in mind? See https://github.com/cldf/pyigt/blob/57dd7900f9cc3a527da19768e2808df3e358bbcd/tests/test_lgr.py for some examples.

File: tests/test_lgr.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    DEM-LOC 1sPOSS-house-PL 3ns-be.NPST
+    'Here are my houses.''
+    """
+
+
+def test_rule6():
+    igt1 = assert_is_valid('puer', 'boy[NOM.SG]')
+    igt2 = assert_is_valid('puer-Ø', 'boy-NOM.SG')
+    assert igt2.primary_text == 'puer'
+    assert igt1 == igt2
```
Comment (xrotwang) [inline tests/test_lgr.py:138]: @fmatter I have no idea how common these non-overt elements are. But what the code does now, is normalize to the format with `Ø` in object language. Then, glosses can be compared for equality - which seems nice.

**Comment by fmatter** [comment]
> @fmatter Does this look like what you had in mind? See https://github.com/cldf/pyigt/blob/57dd7900f9cc3a527da19768e2808df3e358bbcd/tests/test_lgr.py for some examples.

I was originally talking about pycldf -- since you validate word and word gloss correspondences, would you also want to validate morpheme-gloss correspondences there? And I was just thinking of the existing [rudimentary gloss-checking functionality](https://github.com/cldf/pyigt/blob/413d955ad662aa152c11d7fcf3b45c48b36d57b3/src/pyigt/igt.py#L391-L407) in pyigt. But this is of course all the better, whether used for validating CLDF or elsewhere! :)

**Comment by xrotwang** [comment]
I'd be afraid of including this in pycldf. From my experience, expecting word-by-word alignment is already quite ambitious for e.g. sets of examples accompanying typological datasets. Any effort or analysis that would require - say - concordances of gloss labels and morpheme equivalents would need to do more data scrubbing anyway, and could resort to using `pyigt` then.

**Comment by fmatter** [comment]
> I'd be afraid of including this in pycldf. From my experience, expecting word-by-word alignment is already quite ambitious for e.g. sets of examples accompanying typological datasets. Any effort or analysis that would require - say - concordances of gloss labels and morpheme equivalents would need to do more data scrubbing anyway, and could resort to using `pyigt` then.

Yes, that's a reasonable approach, and the more verbose feedback from pyigt will greatly facilitate data scrubbing.

File: tests/test_lgr.py (hunk 1 above)
Comment (LinguList) [reply tests/test_lgr.py:140]: How secure can we be that linguists use the empty set symbol instead of the Ö-Symbol? This is a perfect confusable, and although they should know, linguists typically don't know their Unicode well enough here. So I'd try and have a check for the symbol, and correct it internally (as we do with `:` as a length marker in [CLTS](https://clts.clld.org)), or through some explicit error, or allow for both cases.

Comment (xrotwang) [reply tests/test_lgr.py:140]: Ah, good catch! Will adjust the symbol! I guess, pretending to be able to detect gloss equivalence beyond simple string equality, would again be too ambitious. So I'll skip this normalization.

Comment (LinguList) [reply tests/test_lgr.py]: How awful: I just saw that the null-morpheme is IN the original text, then it does not make sense to normalize, as we cannot tell what we'll get in the original text (although it is unlikely to find an Ø that represents a sound in this form). But then one might want to mention this somewhere, that Ø would not be allowed here.

**Comment by xrotwang** [comment]
I don't really follow. Is the problem that this collides with the proposed
any-sound marker?

Johann-Mattis List ***@***.***> schrieb am Do., 3. Feb. 2022,
19:33:

**Comment by LinguList** [comment]
What I mean is the following (independent of CLTS any soudn etc): if the Ø were in the gloss-text, which is usually English, and rarely Norwegian, you could say: let us normalize Ø to the empty set symbol as a preprocessing step (similarly, as we normalize `:` as the length marker, since we know people notoriously do it wrong). But if it is in the source-language text, it could be that we have some weird transcription where Ø really means something that is not a null-morpheme, so normalization might create some ambiguity.

But it is not that important, as it was not contradicting any decision, rather confirming to NOT try to normalize data, but to add some documentation that Ø cannot pass as ∅.

**Comment by xrotwang** [comment]
Oh, I didn't mean to normalize the symbol, but rather normalize the two
options to mark non-overt element glosses. So I'd consider the Norse symbol
in LGR a simple bug- although the fact of this bug being in the official
documents might well have spread it widely.

Johann-Mattis List ***@***.***> schrieb am Do., 3. Feb. 2022,
19:44:

**Comment by xrotwang** [comment]
@fmatter btw.: I think I'm going to use the LGR as toy example to play around with the idea of some sort of "markdown+CLDF to write linguistic documents" again. I'll do that over at https://github.com/cldf-datasets/lgr - let me know if you want to join.

**Comment by fmatter** [comment]
> @fmatter btw.: I think I'm going to use the LGR as toy example to play around with the idea of some sort of "markdown+CLDF to write linguistic documents" again. I'll do that over at https://github.com/cldf-datasets/lgr - let me know if you want to join.

Absolutely! I would love a proper implementation of a markdown processor for common elements in linguistic texts -- you might remember the ad-hoc markdown pre-processing I used for cariban.clld.org.

**Comment by xrotwang** [comment]
@fmatter So I looked into parsing LGR morphemes properly, and managed to get something working in so far as I can round-trip all morphemes in LGR examples. This is how it looks like:
```python
gw = GlossedWord('yerak~rak-im', 'green~ATT-M.PL')
[(m, m.gloss_elements) for m in gw.word_morphemes]
[(<Morpheme "yerak">, [<GlossElement "yerak">]), (<MorphologicallyBoundWord "rak">, [<GlossElement "rak">]), (<Morpheme "im">, [<GlossElement "im">])]
[(m, m.gloss_elements) for m in gw.gloss_morphemes]
[(<Morpheme "green">, [<GlossElement "green">]), (<MorphologicallyBoundWord "ATT">, [<GlossElement "ATT">]), (<Morpheme "M.PL">, [<GlossElement "M">, <GlossElement "PL">])]
```

It simply parses the stuff left-to-right and determines the class of the next element by the separator used before it. So no attempt is made to figure out which one is the clitic, etc. But each gloss element has a link to previous and next, so some kind of heuristics could be bolted on - maybe
```python
def is_clitic(self):
    return isinstance(self, Clitic) or (isinstance(self.next, Clitic) and len(self) < len(self.next))
```

**Comment by fmatter** [comment]
I was going to look at extracting word forms from IGT this afternoon, so this is very welcome! Will try on a dataset.

**Comment by xrotwang** [comment]
This will be interesting. Since there's quite a few "reserved" characters in LGR, I wouldn't be surprised if this fails badly with real-world data.

**Comment by fmatter** [comment]
First of all, this is NICE!
So far I've only tested it with the LGR examples, and here's some thoughts, questions, and issues:

1. right now there are two implemented ways of segmenting a line, right? There's `igt.phrase_segmented` (and `igt.phrase_segmented2`), as well as the new `GlossedWord`. I focused on the latter, but it seems that functionality it is lacking already exists in `phrase_segmented`.
2. one such functionality is stripping punctuation; if the glossed object line contains punctuation, that should be stripped (probably in `MorphemeList.from_string()`?)
3. I am not fully across how separators are treated in `GlossedWord` -- every morpheme seems to have a `sep`, so in *guny-bi-yarluga*, even the root *yarluga* has `-`. Do these `sep` have a position (pre/post)?
4. There is still the issue of what "Word" means in `GlossedWord`... I think most people (e.g. historical linguists) would be primarily interested in the morphosyntactic word, not prosodic units. This means two things:
    1. morphologically bound but prosodically free elements (*a-nii **-láay***), designated by a space and a - (in either order) should be conflated in a `GlossedWord`: *a-nii**-láay***. This is correctly implemented in that *a-nii -láay* becomes a single `GlossedWord`, but the morpheme *nii* is ` `-final. That space does not belong to the morpheme, but designates a prosodic boundary inside the `GlossedWord`. If round-tripping is desired, then that information would have to be encoded somehow.
    2. if clitics are treated analogously to the "anticlitics" above, then they should become their own `GlossedWord`, so *palasi=lu* 'priest=AND' should become *palasi* 'priest' and *lu* 'and'. The separator then just means that there is no prosodic boundary, which would have to be stored and rendered as `=` instead of ` `. OTOH, if clitics are a special kind of morpheme, then the current implementation works. Maybe a switch like `clitics_as_words` would be nice?

**Comment by fmatter** [comment]
Gloss separation is also nice, but not all `in_gloss_only` separators (`;\:._>[]()`) can be processed equally easily further down the line, because they have really rather different semantics:
    * you can't really do anything with `:`, as segmentation is deliberately obfuscated
    * `_` is also rather pointless, except maybe an extracted meaning should show ` ` instead ('come out')
    * `;` assigns multiple meanings to a form, e.g. *-arum* is both GEN and PL. Separating GEN and PL would matter at the point where a morph-list is created, and it would necessitate a model where a morph can have multiple meanings simultaneously.
    * morphophonological change is **very** language-specific and hard to implement. In a more beefy morphological model, I would probably create an object `MorphoProcess` that can be assigned a meaning, and could optionally be connected to a user-provided algorithm responsible for the transformation of base-form to `\`-glossed form (in this particular case, the `MorphoProcess` *Umlaut* has the meaning 'PL' and transforms *Vater* to *Väter*)
    * `>` assigns roles to persons; parsing this would require modeling some version of A and P, which opens a whole can of worms (modeling predicates, person...)
    * non-overt elements would require a word form being able to say "I carry hidden information about X"
    * to represent inherent categories, morphs would have to be able to carry additional properties apart from form and meaning

**Comment by xrotwang** [comment]
So my first goal with this implementation was being able to fully round-trip all LGR example words/glosses, i.e. parse everything into meaningful objects and then recreate the strings programmatically (i.e. not by looking up the stored original string). I think this is already responsible for one of your issues, the punctuation in the object language words. To recover `sekarang.`, I need to keep the `.` as part of the word. But I think this could be handled with a `@property` `plain_morph` or similar.

The `sep` attribute on `Morpheme` subclasses is a **class** attribute. It's basically used to pick the right class to instantiate a morpheme upon parsing. This parsing happens left-to-right. So the first morpheme is **always** a plain `Morpheme`, the one following a `=` a `Clitic`, etc. This means I can reconstruct the exact input string just by looking at position and class. But as you observe, this may not match expected semantics.

**Comment by xrotwang** [comment]
Regarding you comment on gloss separation: The most trivial processing of the parsed `GlossElement`s would be creating a gloss abbreviation dictionary or similar. So I guess, splitting these does make some sense.

**Comment by xrotwang** [comment]
Regarding the root/prefix/suffix semantics: I think the advantage of the current implementation is that
- it's fairly simple
- it provides enough structure to make it possible to implement some heuristics to guess this
  - somewhere downstream (possibly using language-specific extra information)
  - or in methods on the objects (if reasonable defaults can be constructed)

**Comment by fmatter** [comment]
> So my first goal with this implementation was being able to fully round-trip all LGR example words/glosses, i.e. parse everything into meaningful objects and then recreate the strings programmatically (i.e. not by looking up the stored original string). I think this is already responsible for one of your issues, the punctuation in the object language words. To recover `sekarang.`, I need to keep the `.` as part of the word. But I think this could be handled with a `@property` `plain_morph` or similar.

I see. Or maybe treat punctuation as elements of their own? `[GlossedWord blabla, GlossedWord blabla, Punctuation .]`? This is how punctuation is treated in (XML-)corpora of written language, AFAIK.

> 
> The `sep` attribute on `Morpheme` subclasses is a **class** attribute. It's basically used to pick the right class to instantiate a morpheme upon parsing. This parsing happens left-to-right. So the first morpheme is **always** a plain `Morpheme`, the one following a `=` a `Clitic`, etc. This means I can reconstruct the exact input string just by looking at position and class. But as you observe, this may not match expected semantics.

And the first morpheme has a `sep`, too, right?

**Comment by xrotwang** [comment]
If we do end up adding methods/properties like `is_root` or similar, it may be better to rename the classes to something less confusing, e.g. `Clitic` to `MorphemeAfterEquals`, or `Reduplication` to `MorphemeAfterTilde`.

**Comment by xrotwang** [comment]
Treating punctuation as separate object would seem overkill to me. Stripping it to get a "plain morph" should be simple enough. Treating it as separate object might make some sense, if we want to analyse punctuation as well - e.g. look at all morphemes that can appear before question marks. But from my experience, very few IGT collections will have consistently coded punctuation.

**Comment by fmatter** [comment]
> If we do end up adding methods/properties like `is_root` or similar, it may be better to rename the classes to something less confusing, e.g. `Clitic` to `MorphemeAfterEquals`, or `Reduplication` to `MorphemeAfterTilde`.

This would be more explicit about the internal logic, yes.

I will give another dataset a test drive now, and think about correctly distinguishing prosodic and morphosyntactic boundaries.

**Comment by xrotwang** [comment]
I guess we might need methods `prosodic_words` and `morphosyntactic_words` on the `IGT` level, rather than shoehorning these two aspects into one `GlossedWord`?

**Comment by fmatter** [comment]
Some missing unicode support?

> GlossedWord(word='i-wəi-hpə', gloss='3-old.sister-P.pst', word_morphemes=[<Morpheme "i">, <Morpheme "w?i">, <Morpheme "hp?">], gloss_morphemes=[<Morpheme "3">, <Morpheme "old.sister">, <Morpheme "P.pst">])

**Comment by xrotwang** [comment]
That's intended. repr is supposed to return ascii only.

Florian Matter ***@***.***> schrieb am Do., 17. Feb. 2022,
20:57:

**Comment by xrotwang** [comment]
Now a `GlossedWord` is basically a list of `GlossedMorpheme`s, and functionality like `lexical_concepts` attached to `GlossedMorpheme`, e.g.
```python
from pyigt import GlossedWord
gw = GlossedWord('puer', 'boy[SG.NOM]')
gw[0].lexical_concepts
```

**Comment by xrotwang** [comment]
@fmatter @LinguList should we merge this as it is now?

**Comment by LinguList** [comment]
Yes, fine by me!

**Comment by fmatter** [comment]
It all looks very good -- I haven't given it a proper test drive yet, but will do so soon (converting a text glossed with toolbox to CONLLU).
